### PR TITLE
media/voice: SpeechDetector implementation for the hardware KD & EPD

### PR DESCRIPTION
**Pull request by gcjjyy**
Add logic for initialization of Hardware KeywordDetector / EndPointDetector

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [56a6a6b122dc5f9411003a11db77d072c92f6db3] - Code Rule Check Result:
framework/src/media/voice/HardwareKeywordDetector.h:28: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SoftwareKeywordDetector.h:30: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:32: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [56a6a6b122dc5f9411003a11db77d072c92f6db3] - Code Rule Check (C++) OK.

File: framework/src/media/voice/SpeechDetector.cpp [hunk 1]
```diff
@@ -35,46 +37,81 @@ SpeechDetectorInterface *SpeechDetectorInterface::instance()
...
-	mEndPointDetector->init(samprate, channels);
-
-	return true;
+	return mEndPointDetector->init(samprate, channels);
 }
 
 void SpeechDetector::deinitKeywordDetect()
 {
+	mKeywordDetectorCard = mKeywordDetectorDevice = -1;
+	mKeywordDetector->deinit();
```
Comment (junmin-kim) [inline framework/src/media/voice/SpeechDetector.cpp]: if user do deinitKeywordDetect() twice, data abort will occur.
nullptr check is needed.
```
if (mKeywordDetector) {
    mKeywordDetector->deinit();
}
```

Comment (gcjjyy) [reply framework/src/media/voice/SpeechDetector.cpp]: You're right. I'll fix.

**Comment by seinfra** [comment]
Target : [4c65cc5f5708bc40a095b2d1c59c3bc0f3734b8a] - Code Rule Check Result:
framework/src/media/voice/HardwareKeywordDetector.h:28: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SoftwareKeywordDetector.h:30: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:32: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [4c65cc5f5708bc40a095b2d1c59c3bc0f3734b8a] - Code Rule Check (C++) OK.

File: framework/src/media/voice/SpeechDetector.cpp [hunk 2]
```diff
@@ -35,46 +37,85 @@ SpeechDetectorInterface *SpeechDetectorInterface::instance()
...
+
+	if (result == AUDIO_MANAGER_SUCCESS) {
+		meddbg("KeywordDetector : card_id : %d device_id %d\n", mKeywordDetectorCard, mKeywordDetectorDevice);
+
+		result = change_stream_in_device(mKeywordDetectorCard, mKeywordDetectorDevice);
+		if (result == AUDIO_MANAGER_SUCCESS) {
+			mKeywordDetector = std::make_shared<HardwareKeywordDetector>();
+			return true;
+		} else {
+			return false;
```
Comment (Taejun-Kwon) [inline framework/src/media/voice/SpeechDetector.cpp]: log !

File: framework/src/media/voice/SpeechDetector.cpp [hunk 3]
```diff
@@ -35,46 +37,85 @@ SpeechDetectorInterface *SpeechDetectorInterface::instance()
...
+
+	if (result == AUDIO_MANAGER_SUCCESS) {
+		meddbg("EndPointDetector : card_id : %d device_id %d\n", mEndPointDetectorCard, mEndPointDetectorDevice);
+
+		result = change_stream_in_device(mEndPointDetectorCard, mEndPointDetectorDevice);
+		if (result == AUDIO_MANAGER_SUCCESS) {
+			mEndPointDetector = std::make_shared<HardwareEndPointDetector>();
+			return true;
+		} else {
+			return false;
```
Comment (Taejun-Kwon) [inline framework/src/media/voice/SpeechDetector.cpp]: same as above.

File: framework/src/media/voice/SpeechDetector.cpp (hunk 2 above)
Comment (gcjjyy) [reply framework/src/media/voice/SpeechDetector.cpp]: Fixed

File: framework/src/media/voice/SpeechDetector.cpp (hunk 3 above)
Comment (gcjjyy) [reply framework/src/media/voice/SpeechDetector.cpp]: Fixed

**Comment by seinfra** [comment]
Target : [b19cb1420722a4b4caefce88f4e46cca320bc288] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [b19cb1420722a4b4caefce88f4e46cca320bc288] - Code Rule Check Result:
framework/src/media/voice/HardwareKeywordDetector.h:28: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SoftwareKeywordDetector.h:30: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:32: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [9bdae9990dfb347fb576fb5d2f15b69a1ed2cea0] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [9bdae9990dfb347fb576fb5d2f15b69a1ed2cea0] - Code Rule Check Result:
framework/src/media/voice/HardwareKeywordDetector.h:28: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SoftwareKeywordDetector.h:30: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:32: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [7138fff66097338d0525e70fb11a316cd3e146b4] - Code Rule Check Result:
framework/src/media/voice/HardwareKeywordDetector.h:28: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SoftwareKeywordDetector.h:30: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:32: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [7138fff66097338d0525e70fb11a316cd3e146b4] - Code Rule Check (C++) OK.

File: framework/src/media/voice/SpeechDetector.cpp [hunk 4]
```diff
@@ -35,46 +37,87 @@ SpeechDetectorInterface *SpeechDetectorInterface::instance()
...
+{
+	mKeywordDetectorCard = mKeywordDetectorDevice = -1;
+
+	audio_manager_result_t result = find_stream_in_device_with_process_type(
+		AUDIO_DEVICE_PROCESS_TYPE_SPEECH_DETECTOR,
+		AUDIO_DEVICE_SPEECH_DETECT_KD,
+		&mKeywordDetectorCard, &mKeywordDetectorDevice);
+
+	if (result == AUDIO_MANAGER_SUCCESS) {
+		meddbg("KeywordDetector : card_id : %d device_id %d\n", mKeywordDetectorCard, mKeywordDetectorDevice);
```
Comment (chanijjani) [inline framework/src/media/voice/SpeechDetector.cpp]: I think medvdbg is more appropriate.

File: framework/src/media/voice/SpeechDetector.cpp [hunk 5]
```diff
@@ -35,46 +37,87 @@ SpeechDetectorInterface *SpeechDetectorInterface::instance()
...
-		mEndPointDetector = std::make_shared<HardwareEndPointDetector>();
+	mEndPointDetectorCard = mEndPointDetectorDevice = -1;
+
+	audio_manager_result_t result = find_stream_in_device_with_process_type(
+		AUDIO_DEVICE_PROCESS_TYPE_SPEECH_DETECTOR,
+		AUDIO_DEVICE_SPEECH_DETECT_EPD,
+		&mEndPointDetectorCard, &mEndPointDetectorDevice);
+
+	if (result == AUDIO_MANAGER_SUCCESS) {
+		meddbg("EndPointDetector : card_id : %d device_id %d\n", mEndPointDetectorCard, mEndPointDetectorDevice);
```
Comment (chanijjani) [inline framework/src/media/voice/SpeechDetector.cpp]: I think medvdbg is more appropriate here.

File: framework/src/media/voice/SpeechDetector.cpp (hunk 4 above)
Comment (gcjjyy) [reply framework/src/media/voice/SpeechDetector.cpp]: You're right. I'll fix.

File: framework/src/media/voice/SpeechDetector.cpp (hunk 5 above)
Comment (gcjjyy) [reply framework/src/media/voice/SpeechDetector.cpp]: You're right. I'll fix.

**Comment by seinfra** [comment]
Target : [08f35c6b6688a1911fcbde675898433c8b764b3f] - Code Rule Check Result:
framework/src/media/voice/HardwareKeywordDetector.h:28: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SoftwareKeywordDetector.h:30: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:32: ERROR: [SPC_M_OPR] spaces prohibited around that ':' (ctx:WxW)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:54: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:OxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '<' (ctx:VxV)
framework/src/media/voice/SpeechDetector.h:55: ERROR: [SPC_M_OPR] spaces required around that '>' (ctx:VxW)

**Comment by seinfra** [comment]
Target : [08f35c6b6688a1911fcbde675898433c8b764b3f] - Code Rule Check (C++) OK.

**Pull request merged**